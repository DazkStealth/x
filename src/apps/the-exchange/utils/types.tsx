import { BridgingProvider, ExchangeOffer } from '@etherspot/prime-sdk/dist/sdk/data';
import { Route } from '@lifi/sdk';

export enum CardPosition {
    SWAP = 'SWAP',
    RECEIVE = 'RECEIVE'
}

export type SwapType = {
    fromAmount: number,
    fromTokenAddress: string,
    fromChainId: number;
    fromTokenDecimals: number;
    toTokenAddress: string,
    toChainId: number,
    toTokenDecimals: number;
    slippage: number;
    fromAccountAddress?: string
    provider?: BridgingProvider;
}

export interface SwapOffer {
    tokenAmountToReceive: number;
    offer: Route | ExchangeOffer;
}
