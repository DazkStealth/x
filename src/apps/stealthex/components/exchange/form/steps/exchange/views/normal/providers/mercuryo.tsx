import React, { useEffect } from 'react';
import styled from 'styled-components';
import { ExchangeInfo } from '../../../../../../../../type';

const widgetId = '85c32930-4963-442f-9165-0e213b367633'


const Container = styled.div`
  padding-top: 40px;

  @media (max-width: 934px) {
    padding-top: 20px;
  }
`;

const Mercuryo: React.FC<{ exchangeInfo: ExchangeInfo }> = ({
    exchangeInfo: {
        status,
        currencies,
        currency_from: currencyFrom,
        currency_to: currencyTo,
        address_to: addressTo,
        id,
        amount_from: amountFrom,
        mercuryo_signature: mercuryoSignature,
    },
}) => {
    const from = currencies[currencyFrom];
    const to = currencies[currencyTo];
    const waiting = status == 'waiting';

    useEffect(() => {
        const createWidget = () => {
            const upperCasedFrom = from?.symbol.toUpperCase();
            const upperCasedTo = to?.symbol.toUpperCase();

            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (window as any)?.mercuryoWidget.run({
                widgetId,
                host: document.getElementById('mercuryo-widget'),
                fiatCurrency: upperCasedFrom === 'TL' ? 'TRY' : upperCasedFrom,
                currency: upperCasedTo === 'USDTERC20' ? 'USDT' : upperCasedTo,
                theme: 'stealthex',
                address: addressTo,
                signature: mercuryoSignature,
                merchantTransactionId: id,
                fiatAmount: amountFrom,
                fixAmount: true,
                fixCurrency: true,
                fixFiatAmount: true,
                fixFiatCurrency: true,
                hideAddress: true,
                type: 'buy',
                lang: 'en',
            });
        };

        if ('mercuryoWidget' in window) {
            void createWidget();
        } else {
            const script = document.createElement('script');
            script.async = true;
            script.src = 'https://widget.mercuryo.io/embed.2.0.js';
            script.id = 'mercuryo';
            script.onload = createWidget;
            document.body.appendChild(script);
        }
    }, [
        waiting,
        id,
        from,
        to,
        addressTo,
        amountFrom,
        mercuryoSignature,
    ]);

    return (
        <Container>
            <div id="mercuryo-widget" />
        </Container>
    );
};

export default Mercuryo;
