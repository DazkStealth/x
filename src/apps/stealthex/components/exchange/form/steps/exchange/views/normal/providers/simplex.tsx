import React, { useEffect, useRef } from 'react';
import { ExchangeInfo } from '../../../../../../../../type';


const Simplex: React.FC<{ exchangeInfo: ExchangeInfo }> = ({
    exchangeInfo,
}) => {
    const scriptLoadingRef = useRef(false);

    useEffect(() => {
        if (scriptLoadingRef.current) {
            return;
        }

        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const windowAny = window as any;

        const createWidget = () => {
            windowAny.Simplex.load({
                payment_id: exchangeInfo.payment_id,
            });
};

if ('Simplex' in window) {
    createWidget();
} else {
    scriptLoadingRef.current = true;

    windowAny.simplexAsyncFunction = () => {
        windowAny.Simplex.init({
            public_key: 'pk_live_732a780f-cf93-47d1-ba9b-0494b677ef4b',
        });
    };
    const script = document.createElement('script');
    script.async = true;
    script.src = 'https://cdn.simplex.com/sdk/v1/js/sdk.js';
    script.id = 'simplex';
    script.onload = () => {
        scriptLoadingRef.current = false;
        createWidget();
    };
    document.body.appendChild(script);
}

return () => {
    if ('Simplex' in window) {
        windowAny.Simplex.unload();
    }
};
    }, [exchangeInfo.payment_id, exchangeInfo.id]);

return (
    <form id="simplex-form">
        <div id="checkout-element"></div>
    </form>
);
};

export default Simplex;
