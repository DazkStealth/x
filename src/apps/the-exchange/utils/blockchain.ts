import { BigNumber, BigNumberish } from 'ethers';
import { formatUnits } from 'ethers/lib/utils';
import { formatEther } from 'viem';

export const processBigNumber = (val: BigNumber): number =>
    Number(val.toString());

export const processEth = (val: BigNumberish, dec: number): number => {
    if (typeof val === 'bigint') {
      return +parseFloat(formatEther(val)).toFixed(2);
    }
  
    return +parseFloat(formatUnits(val as BigNumberish, dec));
  };