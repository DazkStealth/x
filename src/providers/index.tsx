import React from 'react';

import LanguageProvider from './LanguageProvider';

const Providers = ({ children }: { children: React.ReactNode }) => {
  return (
    <LanguageProvider>
      {children}
    </LanguageProvider>
  );
}

export default Providers;
