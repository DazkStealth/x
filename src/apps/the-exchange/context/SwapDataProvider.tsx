import React, { createContext, useContext, useState, useMemo } from 'react';
import { Token } from '@etherspot/prime-sdk/dist/sdk/data';

// types
import { SwapOffer } from '../utils/types';

type AmountType = {
  tokenAmount: number;
  usdAmount: number;
}

type ChainType = {
  chainId: number;
  chainName: string;
}

interface SwapDataContextType {
  swapTokenData: Token[];
  setSwapTokenData: React.Dispatch<React.SetStateAction<Token[]>>;
  receiveTokenData: Token[];
  setReceiveTokenData: React.Dispatch<React.SetStateAction<Token[]>>;
  isSwapOpen: boolean;
  setIsSwapOpen: React.Dispatch<React.SetStateAction<boolean>>;
  isReceiveOpen: boolean;
  setIsReceiveOpen: React.Dispatch<React.SetStateAction<boolean>>;
  swapChain: ChainType | undefined;
  setSwapChain: React.Dispatch<React.SetStateAction<ChainType | undefined>>;
  receiveChain: ChainType | undefined;
  setReceiveChain: React.Dispatch<React.SetStateAction<ChainType | undefined>>;
  swapToken: Token | undefined;
  setSwapToken: React.Dispatch<React.SetStateAction<Token | undefined>>;
  receiveToken: Token | undefined;
  setReceiveToken: React.Dispatch<React.SetStateAction<Token | undefined>>;
  amountSwap: AmountType | undefined;
  setAmountSwap: React.Dispatch<React.SetStateAction<AmountType | undefined>>;
  amountReceive: AmountType | undefined;
  setAmountReceive: React.Dispatch<React.SetStateAction<AmountType | undefined>>;
  bestOffer: SwapOffer | undefined;
  setBestOffer: React.Dispatch<React.SetStateAction<SwapOffer | undefined>>;
  searchTokenResult: Token[];
  setSearchTokenResult: React.Dispatch<React.SetStateAction<Token[]>>;

}

export const SwapDataContext = createContext<SwapDataContextType>({
  swapTokenData: [],
  setSwapTokenData: () => {},
  receiveTokenData: [],
  setReceiveTokenData: () => {},
  isSwapOpen: false,
  setIsSwapOpen: () => {},
  isReceiveOpen: false,
  setIsReceiveOpen: () => {},
  swapChain: undefined,
  setSwapChain: () => {},
  receiveChain: undefined,
  setReceiveChain: () => {},
  swapToken: undefined,
  setSwapToken: () => {},
  receiveToken: undefined,
  setReceiveToken: () => {},
  amountSwap: undefined,
  setAmountSwap: () => {},
  amountReceive: undefined,
  setAmountReceive: () => {},
  bestOffer: undefined,
  setBestOffer: () => {},
  searchTokenResult: [],
  setSearchTokenResult: () => {},
});

export const SwapDataProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [swapTokenData, setSwapTokenData] = useState<Token[]>([]);
  const [receiveTokenData, setReceiveTokenData] = useState<Token[]>([]);
  const [isSwapOpen, setIsSwapOpen] = useState<boolean>(false);
  const [isReceiveOpen, setIsReceiveOpen] = useState<boolean>(false);
  const [swapChain, setSwapChain] = useState<ChainType | undefined>(undefined);
  const [receiveChain, setReceiveChain] = useState<ChainType | undefined>(undefined);
  const [swapToken, setSwapToken] = useState<Token | undefined>(undefined);
  const [receiveToken, setReceiveToken] = useState<Token | undefined>(undefined);
  const [amountSwap, setAmountSwap] = useState<AmountType | undefined>(undefined);
  const [amountReceive, setAmountReceive] = useState<AmountType | undefined>(undefined);
  const [bestOffer, setBestOffer] = useState<SwapOffer | undefined>(undefined)
  const [searchTokenResult, setSearchTokenResult] = useState<Token[]>([]);

  const contextValue = useMemo(() => ({
    swapTokenData,
    setSwapTokenData,
    receiveTokenData,
    setReceiveTokenData,
    isSwapOpen,
    setIsSwapOpen,
    isReceiveOpen,
    setIsReceiveOpen,
    swapChain,
    setSwapChain,
    receiveChain,
    setReceiveChain,
    swapToken,
    setSwapToken,
    receiveToken,
    setReceiveToken,
    amountSwap,
    setAmountSwap,
    amountReceive,
    setAmountReceive,
    bestOffer,
    setBestOffer,
    searchTokenResult,
    setSearchTokenResult,
  }), [
    swapTokenData,
    receiveTokenData,
    isSwapOpen,
    isReceiveOpen,
    swapChain,
    receiveChain,
    swapToken,
    receiveToken,
    amountSwap,
    amountReceive,
    bestOffer,
    searchTokenResult,
  ]);

  return (
    <SwapDataContext.Provider value={contextValue}>
      {children}
    </SwapDataContext.Provider>
  );
};

export const useSwapDataContext = () => useContext(SwapDataContext);
