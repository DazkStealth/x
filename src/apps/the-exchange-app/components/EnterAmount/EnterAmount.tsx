import BodySmall from '../Typography/BodySmall';
import SendArrow from '../../images/send-arrow.png';
import ReceiveArrow from '../../images/receive-arrow.png';
import { useContext, useState, useEffect } from 'react';
import { SwapReceive } from '../../utils/types';
import { SwapDataContext } from '../../context/SwapDataProvider';
import { useEtherspotPrices } from '@etherspot/transaction-kit';
import useOffer from '../../hooks/useOffer';

type EnterAmountProps = {
    type: SwapReceive;
    tokenSymbol?: string;
}

const EnterAmount = ({ type, tokenSymbol }: EnterAmountProps) => {
    const [isHover, setIsHover] = useState<boolean>(false);
    const [isFocus, setIsFocus] = useState<boolean>(false);
    const { amountSwap, setAmountSwap, amountReceive, setAmountReceive, swapToken, receiveToken, setBestOffer } = useContext(SwapDataContext);
    const [inputValue, setInputValue] = useState<string>('0.00');
    const { getPrice } = useEtherspotPrices();
    const { getBestOffer } = useOffer(swapToken?.chainId || 0);

    const getOffer = async () => {
        const params = {
            fromAmount: amountSwap?.tokenAmount || 0,
            fromTokenAddress: swapToken?.address || '',
            fromChainId: swapToken?.chainId || 0,
            fromTokenDecimals: swapToken?.decimals || 0,
            toTokenAddress: receiveToken?.address || '',
            toChainId: receiveToken?.chainId || 0,
            toTokenDecimals: receiveToken?.decimals || 0,
        }
        const offer = await getBestOffer(params);

        return offer;
    }
    
    useEffect(() => {
        getOffer().then((offer) => setBestOffer(offer));
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [amountSwap, swapToken, receiveToken])

    useEffect(() => {
        if (type === SwapReceive.SWAP) {
            setInputValue(amountSwap ? amountSwap.tokenAmount.toString() : '0.00');
        } else {
            setInputValue(amountReceive ? amountReceive.tokenAmount.toString() : '0.00');
        }
    }, [type, amountSwap, amountReceive]);

    const handleChange = async (e: React.ChangeEvent<HTMLInputElement>) => {
        const value = e.target.value;
        setInputValue(value);
    
        if (type === SwapReceive.SWAP && swapToken) {
            const usdPrice = await getPrice(swapToken.address, swapToken.chainId);
            if (usdPrice) {
                setAmountSwap({ tokenAmount: Number(value), usdAmount: usdPrice.usd * Number(value) });
            } else {
                setAmountSwap({ tokenAmount: Number(value), usdAmount: 0 });
            }
        } else if (type === SwapReceive.RECEIVE && receiveToken) {
            const usdPrice = await getPrice(receiveToken.address, receiveToken.chainId);
            if (usdPrice) {
                setAmountReceive({ tokenAmount: Number(value), usdAmount: usdPrice.usd * Number(value) });
            } else {
                setAmountReceive({ tokenAmount: Number(value), usdAmount: 0 });
            }
        }
    };


    return (
        <div 
            onMouseEnter={() => setIsHover(true)} 
            onMouseLeave={() => setIsHover(false)} 
            className="flex flex-col gap-1 group"
        >
            <BodySmall className={`${(isHover && !isFocus) && 'text-black_grey/[.4]'} font-normal`}>
                {tokenSymbol}
            </BodySmall>
            <input 
                type='number' 
                value={inputValue} 
                onChange={handleChange} 
                onFocus={() => setIsFocus(true)} 
                onBlur={() => setIsFocus(false)} 
                className='text-black_grey font-normal !text-3xl outline-none focus:outline-none focus:ring-0 focus:bg-[#292D32]/[.05] focus:border-b focus:border-b-black_grey group-hover:bg-[#292D32]/[.05] group-hover:border-b group-hover:border-b-black_grey mobile:max-w-[120px] tablet:max-w-[260px] desktop:max-w-[260px] xs:max-w-[100px]' 
            />
            <div className='flex justify-between'>
                <BodySmall className={`${(isHover && !isFocus) && 'text-black_grey/[.4]'}`}>
                    ${type === SwapReceive.SWAP ? amountSwap?.usdAmount.toFixed(2) : amountReceive?.usdAmount.toFixed(2)}
                </BodySmall>
                <img src={type === SwapReceive.SWAP ? SendArrow : ReceiveArrow} alt={type === SwapReceive.SWAP ? 'Send' : 'Receive'} />
            </div>
        </div>
    )
}

export default EnterAmount;
