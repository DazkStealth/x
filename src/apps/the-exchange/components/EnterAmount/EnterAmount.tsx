import { useContext, useState, useEffect } from 'react';
import _ from 'lodash';
import { useEtherspotPrices, useEtherspotUtils, useWalletAddress } from '@etherspot/transaction-kit';
import { hasThreeZerosAfterDecimal } from '../../utils/converters';
import { processEth } from '../../utils/blockchain';

// hooks
import useOffer from '../../hooks/useOffer';
import useAccountBalances from '../../../../hooks/useAccountBalances';

// context
import { SwapDataContext } from '../../context/SwapDataProvider';

// types
import { CardPosition } from '../../utils/types';

// components
import BodySmall from '../Typography/BodySmall';
import NumberText from '../Typography/NumberText';
import Body from '../Typography/Body';
import { CircularProgress } from '@mui/material';

// images
import SendArrow from '../../images/send-arrow.png';
import ReceiveArrow from '../../images/receive-arrow.png';

type EnterAmountProps = {
    type: CardPosition;
    tokenSymbol?: string;
};

const EnterAmount = ({ type, tokenSymbol }: EnterAmountProps) => {
    const [isHover, setIsHover] = useState<boolean>(false);
    const [isFocus, setIsFocus] = useState<boolean>(false);
    const {
        amountSwap,
        setAmountSwap,
        amountReceive,
        setAmountReceive,
        swapToken,
        receiveToken,
        setBestOffer,
        bestOffer,
    } = useContext(SwapDataContext);
    const [inputValue, setInputValue] = useState<number>(0);
    const { getPrice } = useEtherspotPrices();
    const balances = useAccountBalances();
    const walletAddress = useWalletAddress();
    const { isZeroAddress, addressesEqual } = useEtherspotUtils();
    const { getBestOffer } = useOffer(swapToken?.chainId || 0);
    const [isOfferLoading, setIsOfferLoading] = useState<boolean>(false);
    const [isNoOffer, setIsNoOffer] = useState<boolean>(false);
    const [errorMessage, setErrorMessage] = useState<string>('');

    // Gets the best swap offer
    const getOffer = async () => {
        const params = {
            fromAmount: amountSwap?.tokenAmount || 0,
            fromTokenAddress: swapToken?.address || '',
            fromChainId: swapToken?.chainId || 0,
            fromTokenDecimals: swapToken?.decimals || 0,
            toTokenAddress: receiveToken?.address || '',
            toChainId: receiveToken?.chainId || 0,
            toTokenDecimals: receiveToken?.decimals || 0,
            slippage: 0.05,
        };

        try {
            const offer = await getBestOffer(params);

            if (offer && receiveToken) {
                const usdPrice = await getPrice(receiveToken.address, receiveToken.chainId);
                if (usdPrice) {
                    setAmountReceive({
                        tokenAmount: offer.tokenAmountToReceive,
                        usdAmount: usdPrice.usd * offer.tokenAmountToReceive,
                    });
                } else {
                    setAmountReceive({ tokenAmount: offer.tokenAmountToReceive, usdAmount: 0 });
                }
            } else {
                setIsNoOffer(true);
            }

            return offer;
        } catch (error) {
            console.error('Failed to fetch the best offer');
            return null;
        } finally {
            setIsOfferLoading(false);
        }
    };

    // Debounced getOffer function with 2-second delay
    const debouncedGetOffer = _.debounce(() => {
        getOffer().then((offer) => {
            if (offer) setBestOffer(offer);
        });
    }, 2000);

    // Similar function being used for the SendModal and AccountModal to check tokens balances
    const tokenBalanceLimit = (tokenAmount: number) => {
        if (!swapToken) return;

        const assetBalance =
            swapToken.chainId &&
            balances[swapToken.chainId]?.[walletAddress as string]?.find((balance) => {
                if (!swapToken.address) {
                    setErrorMessage('This token is not being supported by your wallet');
                    return;
                }

                const assetAddress = swapToken.address;
                const isNativeBalance = balance.token === null || isZeroAddress(balance.token);

                return (
                    (isNativeBalance && isZeroAddress(assetAddress)) ||
                    addressesEqual(balance.token, assetAddress)
                );
            });

        const assetBalanceValue = assetBalance ? assetBalance.balance : '0';
        const balance = processEth(assetBalanceValue, swapToken.decimals ?? 18);

        // Check if the value exceeds the max token amount limit
        if (balance && tokenAmount > balance) {
            setErrorMessage(`The max amount limit in your wallet is ${balance.toFixed(4)} ${swapToken?.symbol}`);
        }
    };

    // getOffer will be called every time the swap amount or the swap/receive token is changed
    useEffect(() => {
        setInputValue(amountSwap ? amountSwap.tokenAmount : 0);
        if (amountSwap?.tokenAmount) {
            setIsOfferLoading(true);
            debouncedGetOffer();
            tokenBalanceLimit(amountSwap?.tokenAmount);
        }
        // Clean-up debounce on component unmount
        return () => {
            debouncedGetOffer.cancel();
        };
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [amountSwap, swapToken, receiveToken]);

    // When the token amount value changes, the input value reflects that change
    // the handleTokenAmountChange will make sure that we get a USD price, and will then add it to amountSwap
    const handleTokenAmountChange = async (e: React.ChangeEvent<HTMLInputElement>) => {
        const value = e.target.value;
        setInputValue(Number(value));
        setErrorMessage('');

        // Might need to find a way to debounce the getPrice call too
        if (type === CardPosition.SWAP && swapToken) {
            const usdPrice = await getPrice(swapToken.address, swapToken.chainId);
            if (usdPrice) {
                setAmountSwap({ tokenAmount: Number(value), usdAmount: usdPrice.usd * Number(value) });
            } else {
                setAmountSwap({ tokenAmount: Number(value), usdAmount: 0 });
            }
        }
    };

    // Function to render offer based on loading and offer state
    const renderOffer = () => {
        if (isOfferLoading) {
            return <CircularProgress size={36} sx={{ color: '#343434' }} />;
        } else if (isNoOffer) {
            return <Body>No offer</Body>;
        } else if (bestOffer) {
            return (
                <NumberText className="text-black_grey font-normal text-3xl break-words mobile:max-w-[180px] tablet:max-w-[260px] desktop:max-w-[260px] xs:max-w-[110px]">
                    {hasThreeZerosAfterDecimal(bestOffer?.tokenAmountToReceive)
                        ? bestOffer?.tokenAmountToReceive.toFixed(8)
                        : bestOffer?.tokenAmountToReceive.toFixed(4)}
                </NumberText>
            );
        }
    };

    return (
        <div
            onMouseEnter={() => setIsHover(true)}
            onMouseLeave={() => setIsHover(false)}
            className="flex flex-col gap-1 group"
        >
            <BodySmall className={`${isHover && !isFocus && 'text-black_grey/[.4]'} font-normal`}>
                {tokenSymbol}
            </BodySmall>
            {type === CardPosition.SWAP ? (
                <>
                    <input
                        type="number"
                        value={inputValue}
                        onChange={(e) => handleTokenAmountChange(e)}
                        onFocus={() => setIsFocus(true)}
                        onBlur={() => setIsFocus(false)}
                        style={{ width: '100%' }}
                        className="text-black_grey font-normal !text-3xl outline-none focus:outline-none focus:ring-0 focus:bg-[#292D32]/[.05] focus:border-b focus:border-b-black_grey group-hover:bg-[#292D32]/[.05] group-hover:border-b group-hover:border-b-black_grey"
                    />
                    {errorMessage && <BodySmall>{errorMessage}</BodySmall>}
                </>
            ) : (
                renderOffer()
            )}
            <div className="flex justify-between">
                <BodySmall className={`${isHover && !isFocus && 'text-black_grey/[.4]'}`}>
                    ${type === CardPosition.SWAP ? amountSwap?.usdAmount.toFixed(2) : amountReceive?.usdAmount.toFixed(2)}
                </BodySmall>
                <img src={type === CardPosition.SWAP ? SendArrow : ReceiveArrow} alt={type === CardPosition.SWAP ? 'Send' : 'Receive'} />
            </div>
        </div>
    );
};

export default EnterAmount;
