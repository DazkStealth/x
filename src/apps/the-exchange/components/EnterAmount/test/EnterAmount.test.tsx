import renderer from 'react-test-renderer';
import { render, fireEvent, waitFor } from '@testing-library/react';

// context
import { SwapDataContext } from '../../../context/SwapDataProvider';

// types
import { CardPosition } from '../../../utils/types';
import { AccountBalancesListenerRef } from '../../../../../providers/AccountBalancesProvider';

// components
import EnterAmount from '../EnterAmount';

const mockTokenAssets = [
    { address: '0x01', name: 'Ether', symbol: 'ETH', chainId: 1, decimals: 18, icon: 'iconEth.png' },
    { address: '0x02', name: 'Matic', symbol: 'MATIC', chainId: 137, decimals: 18, icon: 'iconMatic.png' },
  ];
  
  // Mock data and functions
  const mockSetSwapTokenData = jest.fn();
  const mockSetReceiveTokenData = jest.fn();
  const mockSetIsSwapOpen = jest.fn();
  const mockSetIsReceiveOpen = jest.fn();
  const mockSetSwapChain = jest.fn();
  const mockSetReceiveChain = jest.fn();
  const mockSetSwapToken = jest.fn();
  const mockSetReceiveToken = jest.fn();
  const mockSetAmountSwap = jest.fn();
  const mockSetAmountReceive = jest.fn();
  
  // Mock context values
  const mockContextValue = {
    swapTokenData: mockTokenAssets,
    setSwapTokenData: mockSetSwapTokenData,
    receiveTokenData: mockTokenAssets,
    setReceiveTokenData: mockSetReceiveTokenData,
    isSwapOpen: false,
    setIsSwapOpen: mockSetIsSwapOpen,
    isReceiveOpen: false,
    setIsReceiveOpen: mockSetIsReceiveOpen,
    swapChain: { chainId: 1, chainName: 'Ethereum' },
    receiveChain: { chainId: 137, chainName: 'Polygon' },
    swapToken: mockTokenAssets[0],
    receiveToken: mockTokenAssets[1],
    setSwapChain: mockSetSwapChain,
    setReceiveChain: mockSetReceiveChain,
    setSwapToken: mockSetSwapToken,
    setReceiveToken: mockSetReceiveToken,
    amountSwap: { tokenAmount: 0.1, usdAmount: 3000 },
    amountReceive: { tokenAmount: 10, usdAmount: 3000 },
    setAmountSwap: mockSetAmountSwap,
    setAmountReceive: mockSetAmountReceive,
    bestOffer: undefined,
    setBestOffer: jest.fn(),
    searchTokenResult: [],
    setSearchTokenResult: jest.fn(),
  };

  // Mock transaction-kit hooks being used
jest.mock('@etherspot/transaction-kit', () => ({
    __esModule: true,
    useEtherspotSwaps: jest.fn().mockReturnValue({
      getOffers: jest.fn().mockResolvedValue([]),
      prepareCrossChainOfferTransactions: jest.fn().mockResolvedValue({}),
      getQuotes: jest.fn().mockResolvedValue({}),
    }),
    useEtherspotPrices: jest.fn().mockReturnValue({
      getPrice: jest.fn(),
      getPrices: jest.fn(),
    }),
    useWalletAddress: jest.fn().mockReturnValue({
      walletAddress: jest.fn(),
    }),
    useEtherspotUtils: jest.fn().mockReturnValue({
      isZeroAddress: jest.fn(),
      addressesEqual: jest.fn(),
    }),
    useEtherspotBalances: jest.fn().mockReturnValue({
        getAccountBalances: jest.fn(),
      }),
  }));

    // Mock useAccountBalances hook
    jest.mock('../../../../../hooks/useAccountBalances', () => ({
    __esModule: true,
    default: jest.fn().mockReturnValue({
      listenerRef: { current: {} as AccountBalancesListenerRef },
      data: {
        balances: {
          '0x01': { balance: '0.2', usdValue: '6000' },
          '0x02': { balance: '20', usdValue: '6000' },
        },
        updateData: false,
        setUpdateData: jest.fn(),
      },
    }),
  }));

describe('<DropdownTokenList />', () => {  
    it('renders correctly and matches snapshot', () => {
        const tree = renderer
            .create(<EnterAmount type={CardPosition.SWAP} tokenSymbol={mockContextValue.swapToken.symbol} />)
            .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('handles token amount change in Swap card', async () => {
        const { getByTestId } = render(
            <SwapDataContext.Provider value={{ ...mockContextValue, amountSwap: {tokenAmount: 0, usdAmount: 0} } }>
                <EnterAmount type={CardPosition.SWAP} tokenSymbol={mockContextValue.swapToken.symbol} />
            </SwapDataContext.Provider>
        );

        const inputElement = getByTestId('enter-amount-input');
        fireEvent.change(inputElement, { target: { value: '50' } });

        await waitFor(() => {
            expect(mockContextValue.setAmountSwap).toHaveBeenCalledWith({ tokenAmount: 50, usdAmount: 0 });
          });
    });

    it('handles mouse hover behavior', () => {
        const { getByText } = render(
            <SwapDataContext.Provider value={mockContextValue}>
                <EnterAmount type={CardPosition.SWAP} tokenSymbol={mockContextValue.swapToken.symbol} />
            </SwapDataContext.Provider>
        );

        const tokenSymbolElement = getByText(mockContextValue.swapToken.symbol);

        fireEvent.mouseEnter(tokenSymbolElement);
        expect(tokenSymbolElement).toHaveClass('text-black_grey/[.4]');

        fireEvent.mouseLeave(tokenSymbolElement);
        expect(tokenSymbolElement).not.toHaveClass('text-black_grey/[.4]');
    });

});
