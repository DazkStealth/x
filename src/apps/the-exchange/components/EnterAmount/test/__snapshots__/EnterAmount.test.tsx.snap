// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownTokenList /> renders correctly and matches snapshot 1`] = `
<div
  className="flex flex-col gap-1 group"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <p
    className="text-xs font-normal text-black_grey false font-normal"
  >
    ETH
  </p>
  <input
    className="text-black_grey font-normal !text-3xl outline-none focus:outline-none focus:ring-0 focus:bg-[#292D32]/[.05] focus:border-b focus:border-b-black_grey group-hover:bg-[#292D32]/[.05] group-hover:border-b group-hover:border-b-black_grey"
    data-testid="enter-amount-input"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    style={
      {
        "width": "100%",
      }
    }
    type="number"
    value={0}
  />
  <div
    className="flex justify-between"
  >
    <p
      className="text-xs font-normal text-black_grey false"
    >
      $
    </p>
    <img
      alt="Send"
      src="send-arrow.png"
    />
  </div>
</div>
`;
